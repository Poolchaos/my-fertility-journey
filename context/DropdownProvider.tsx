import React, { createContext, useContext, useState, useCallback } from 'react';

interface DropdownContextProps {
  activeDropdown: string | null;
  toggleDropdown: (dropdown: string) => void;
  closeDropdown: () => void;
}

const DropdownContext = createContext<DropdownContextProps | undefined>(
  undefined
);

export const DropdownProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [activeDropdown, setActiveDropdown] = useState<string | null>(null);

  const toggleDropdown = useCallback((dropdown: string) => {
    setActiveDropdown((prev) => (prev === dropdown ? null : dropdown));
  }, []);

  const closeDropdown = useCallback(() => {
    setActiveDropdown(null);
  }, []);

  return (
    <DropdownContext.Provider
      value={{ activeDropdown, toggleDropdown, closeDropdown }}
    >
      {children}
    </DropdownContext.Provider>
  );
};

export const useDropdown = () => {
  const context = useContext(DropdownContext);
  if (!context) {
    throw new Error('useDropdown must be used within a DropdownProvider');
  }
  return context;
};
