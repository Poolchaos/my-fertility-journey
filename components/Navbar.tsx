import React from 'react';
import { View, TouchableOpacity, Text } from 'react-native';
import { FontAwesome5 } from '@expo/vector-icons';
import { UseEnvironment } from '../utils/environmentUtils';
import NotificationDropdown from './NotificationDropdown';
import ProfileDropdown from './ProfileDropdown';
import { getStyle } from './layout/styles';
import { useDropdown } from '../context/DropdownProvider';

FontAwesome5.defaultProps.solid = true;

const Navbar: React.FC = () => {
  const { environment, tailwind } = UseEnvironment();
  const { activeDropdown, toggleDropdown } = useDropdown();

  return (
    <View
      style={[
        tailwind('flex-row justify-end items-center px-4 py-13 z-50'),
        { backgroundColor: getStyle('bg-white') },
      ]}
    >
      {/* Notifications */}
      <View
        style={[
          tailwind('z-50'),
          { position: environment.width < 535 ? 'static' : 'relative' },
        ]}
      >
        <TouchableOpacity
          onPress={() => toggleDropdown('notifications')}
          style={[tailwind('relative mr-4'), { zIndex: 51 }]}
        >
          <FontAwesome5 name="bell" size={15} color="#9d9d9d" />
          {/* Notification Badge */}
          <View
            style={[
              tailwind(
                'absolute -top-3 -right-3 bg-red-500 rounded-full w-18 h-18 flex items-center justify-center'
              ),
              { backgroundColor: '#ff5651' },
            ]}
          >
            <Text
              style={[
                tailwind('text-white text-xxs font-bold'),
                { color: '#ffffff' },
              ]}
            >
              2
            </Text>
          </View>
        </TouchableOpacity>
        {activeDropdown === 'notifications' && <NotificationDropdown />}
      </View>

      {/* Profile Dropdown */}
      <ProfileDropdown />
    </View>
  );
};

export default React.memo(Navbar);
