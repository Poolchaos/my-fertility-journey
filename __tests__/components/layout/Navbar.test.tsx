import React from 'react';
const { View, Text } = require('react-native');
import { render, screen, fireEvent } from '@testing-library/react-native';
import Navbar from '../../../components/layout/Navbar';
import { useDropdown } from '../../../context/DropdownProvider';

jest.mock('../../../context/DropdownProvider', () => ({
  useDropdown: jest.fn(),
}));

jest.mock('@expo/vector-icons', () => ({
  FontAwesome5: (props: {
    name: string;
    size: number;
    color: string;
    solid?: boolean;
  }) => {
    return (
      <View testID={`${props.name}-icon`}>
        <Text>{props.name}</Text>
      </View>
    );
  },
}));

jest.mock('expo-font', () => ({
  loadAsync: jest.fn(),
}));

describe('Navbar Component', () => {
  let mockToggleDropdown: jest.Mock;

  beforeEach(() => {
    jest.clearAllMocks();
    mockToggleDropdown = jest.fn();
    (useDropdown as jest.Mock).mockReturnValue({
      activeDropdown: null,
      toggleDropdown: mockToggleDropdown,
    });

    const mockedFontAwesome5 = require('@expo/vector-icons/FontAwesome5');
    mockedFontAwesome5.defaultProps = { solid: true };
  });

  it('renders Navbar correctly', () => {
    render(<Navbar />);

    expect(screen.getAllByTestId('bell-icon').length).toBeGreaterThan(0);
    expect(screen.getByTestId('profile-icon')).toBeTruthy();
  });

  it('toggles the notification dropdown on bell icon press', () => {
    render(<Navbar />);

    const bellIcons = screen.getAllByTestId('bell-icon');
    const bellIcon = bellIcons[0];

    fireEvent.press(bellIcon);

    expect(mockToggleDropdown).toHaveBeenCalledTimes(1);
    expect(mockToggleDropdown).toHaveBeenCalledWith('notifications');
  });

  it('renders NotificationDropdown when active', () => {
    (useDropdown as jest.Mock).mockReturnValue({
      activeDropdown: 'notifications',
      toggleDropdown: mockToggleDropdown,
    });

    render(<Navbar />);
    expect(screen.getByTestId('notification-dropdown')).toBeTruthy();
  });

  it('toggles the profile dropdown on profile icon press', () => {
    render(<Navbar />);

    const profileIcon = screen.getByTestId('profile-icon');
    fireEvent.press(profileIcon);

    expect(mockToggleDropdown).toHaveBeenCalledTimes(1);
    expect(mockToggleDropdown).toHaveBeenCalledWith('profile');
  });

  it('matches snapshot', () => {
    const { toJSON } = render(<Navbar />);
    expect(toJSON()).toMatchSnapshot();
  });
});
