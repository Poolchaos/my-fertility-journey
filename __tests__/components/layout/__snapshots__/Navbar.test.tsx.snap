// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component matches snapshot 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "paddingBottom": 13,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 13,
        "zIndex": 50,
      },
      {
        "backgroundColor": "rgb(255 255 255)",
      },
    ]
  }
>
  <View
    style={
      [
        {
          "zIndex": 50,
        },
        {
          "position": "relative",
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "marginRight": 16,
          "opacity": 1,
          "position": "relative",
          "zIndex": 51,
        }
      }
      testID="bell-icon"
    >
      <View
        testID="bell-icon"
      >
        <Text>
          bell
        </Text>
      </View>
      <View
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "rgb(239 68 68 / var(--tw-bg-opacity, 1))",
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
              "display": "flex",
              "height": 18,
              "justifyContent": "center",
              "position": "absolute",
              "right": -12,
              "top": -12,
              "width": 18,
            },
            {
              "backgroundColor": "#ff5651",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgb(255 255 255 / var(--tw-text-opacity, 1))",
                "fontSize": 9.6,
                "fontWeight": "700",
              },
              {
                "color": "#ffffff",
              },
            ]
          }
        >
          2
        </Text>
      </View>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="profile-icon"
  >
    <View
      style={
        [
          {
            "display": "flex",
            "flexDirection": "row",
          },
          {
            "position": "static",
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
              "display": "flex",
              "height": 15,
              "justifyContent": "center",
              "marginLeft": 10,
              "marginRight": 10,
              "paddingBottom": 20,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 20,
              "width": 15,
            },
            {
              "backgroundColor": "rgb(243 249 250)",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 12,
                "fontWeight": "700",
                "lineHeight": 28,
              },
              {
                "color": "rgb(87 131 136)",
              },
            ]
          }
        >
          AS
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "opacity": 1,
            "position": "relative",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
              "display": "flex",
              "height": 15,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              {
                "color": "rgb(160 160 160 / var(--tw-text-opacity, 1))",
                "fontSize": 12,
                "fontWeight": "600",
              }
            }
          >
            Adrian Stefan
          </Text>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "width": 30,
            }
          }
        >
          <Image
            source={
              {
                "testUri": "../../../assets/images/caret-down.svg",
              }
            }
            style={
              {
                "height": 22,
                "width": 22,
              }
            }
            tintColor="rgb(103 173 185)"
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
